File: app/src/main/java/com/aaronbrecher/neverlate/adapters/EventListAdapter.java
package com.aaronbrecher.neverlate.adapters;

import android.content.Context;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import com.aaronbrecher.neverlate.Constants;
import com.aaronbrecher.neverlate.Utils.DirectionsUtils;
import com.aaronbrecher.neverlate.databinding.EventsListItemBinding;
import com.aaronbrecher.neverlate.interfaces.ListItemClickListener;
import com.aaronbrecher.neverlate.models.Event;

import org.threeten.bp.LocalDateTime;
import org.threeten.bp.format.DateTimeFormatter;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class EventListAdapter extends RecyclerView.Adapter<EventListAdapter.EventViewHolder> implements Filterable {
    private static final String TAG = EventListAdapter.class.getSimpleName();
    private List<Event> mEvents;
    private List<Event> mFilteredEvents;
    private ListItemClickListener mClickListener;

    public Context getContext() {
        return mContext;
    }

    private Context mContext;

    /**
     * Constructor for the adapter
     *
     * @param events        The list of Events to be used - may be null and set later with swapLists
     * @param clickListener Interface to handle clicks will be used also as context - MUST BE A VALID CONTEXT!!
     */
    public EventListAdapter(List<Event> events,Context context, ListItemClickListener clickListener) {
        mEvents = events;
        mFilteredEvents = events;
        mClickListener = clickListener;
        mContext = context;
    }

    @NonNull
    @Override
    public EventViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        return new EventViewHolder(EventsListItemBinding.inflate(inflater, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final EventViewHolder holder, final int position) {
        Event event = mFilteredEvents.get(position);
        holder.binding.eventTitle.setText(event.getTitle());
        holder.binding.eventLocation.setText(event.getLocation());
        if (event.getDistance() != Constants.ROOM_INVALID_LONG_VALUE) holder.binding.eventDistance.setText(
                DirectionsUtils.getHumanReadableDistance(mContext,event.getDistance(),  PreferenceManager.getDefaultSharedPreferences(mContext)));
        LocalDateTime time = event.getStartTime();
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("h:mm a");
        if (time != null) holder.binding.eventTimeTo.setText(time.format(formatter));
    }

    @Override
    public int getItemCount() {
        if (mFilteredEvents == null) return 0;
        return mFilteredEvents.size();
    }


    public void swapLists(List<Event> events) {
        this.mEvents = new ArrayList<>(events);
        this.mFilteredEvents = mEvents;
        notifyDataSetChanged();
    }

    public void removeAt(int index){
        mFilteredEvents.remove(index);
        notifyItemRemoved(index);
    }

    public void insertAt(int index, Event event){
        try {
            mFilteredEvents.add(index, event);
        }catch (IndexOutOfBoundsException e){
            mFilteredEvents.add(event);
        }
        notifyItemInserted(index);
    }

    public class EventViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private EventsListItemBinding binding;

        public EventViewHolder(EventsListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            Event event = mEvents.get(getAdapterPosition());
            mClickListener.onListItemClick(event);
        }
    }

    /**
     * Filter the list based on the searchbar in menu
     * @return
     */
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                String query = constraint.toString();
                List<Event> filtered = new ArrayList<>();
                if(query.isEmpty()){
                    filtered = mEvents;
                }else{
                    for(Event event : mEvents){
                        if(event.getTitle().toLowerCase().contains(query.toLowerCase())){
                            filtered.add(event);
                        }
                    }
                }
                FilterResults results = new FilterResults();
                results.count = filtered.size();
                results.values = filtered;
                return results;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                mFilteredEvents = (ArrayList<Event>)results.values;
                notifyDataSetChanged();
            }
        };
    }
}

File: app/src/main/java/com/aaronbrecher/neverlate/adapters/EventListAdapter.kt
package com.aaronbrecher.neverlate.adapters

import android.content.Context
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable

import com.aaronbrecher.neverlate.Constants
import com.aaronbrecher.neverlate.Utils.DirectionsUtils
import com.aaronbrecher.neverlate.databinding.EventsListItemBinding
import com.aaronbrecher.neverlate.interfaces.ListItemClickListener
import com.aaronbrecher.neverlate.models.Event

import org.threeten.bp.LocalDateTime
import org.threeten.bp.format.DateTimeFormatter

import java.util.ArrayList

import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView

/**
 * Constructor for the adapter
 *
 * @param mEvents        The list of Events to be used - may be null and set later with swapLists
 * @param mClickListener Interface to handle clicks will be used also as context - MUST BE A VALID CONTEXT!!
 */
class EventListAdapter (private var mEvents: MutableList<Event>?, val context: Context, private val mClickListener: ListItemClickListener) : RecyclerView.Adapter<EventListAdapter.EventViewHolder>(), Filterable {
    private var mFilteredEvents: MutableList<Event>

    init {
        mFilteredEvents = mEvents ?: ArrayList()
    }

    @NonNull
    override fun onCreateViewHolder(@NonNull parent: ViewGroup, viewType: Int): EventViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        return EventViewHolder(EventsListItemBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(@NonNull holder: EventViewHolder, position: Int) {
        val event = mFilteredEvents[position]
        holder.binding.eventTitle.text = event.title
        holder.binding.eventLocation.text = event.location
        if (event.distance != Constants.ROOM_INVALID_LONG_VALUE)
            holder.binding.eventDistance.text = DirectionsUtils.getHumanReadableDistance(context, event.distance?: -1, PreferenceManager.getDefaultSharedPreferences(context))
        val time = event.startTime
        val formatter = DateTimeFormatter.ofPattern("h:mm a")
        if (time != null) holder.binding.eventTimeTo.text = time.format(formatter)
    }

    override fun getItemCount(): Int {
        return mFilteredEvents.size
    }


    fun swapLists(events: List<Event>) {
        this.mEvents = ArrayList(events)
        this.mFilteredEvents = mEvents ?: ArrayList()
        notifyDataSetChanged()
    }

    fun removeAt(index: Int) {
        mFilteredEvents.removeAt(index)
        notifyItemRemoved(index)
    }

    fun insertAt(index: Int, event: Event) {
        try {
            mFilteredEvents.add(index, event)
        } catch (e: IndexOutOfBoundsException) {
            mFilteredEvents.add(event)
        }

        notifyItemInserted(index)
    }

    inner class EventViewHolder(val binding: EventsListItemBinding) : RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val event = mEvents!![adapterPosition]
            mClickListener.onListItemClick(event)
        }
    }

    /**
     * Filter the list based on the searchbar in menu
     * @return
     */
    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): Filter.FilterResults {
                val query = constraint.toString()
                var filtered: MutableList<Event>? = ArrayList()
                if (query.isEmpty()) {
                    filtered = mEvents
                } else {
                    for (event in mEvents!!) {
                        if (event.title!!.toLowerCase().contains(query.toLowerCase())) {
                            filtered!!.add(event)
                        }
                    }
                }
                val results = Filter.FilterResults()
                results.count = filtered!!.size
                results.values = filtered
                return results
            }

            override fun publishResults(constraint: CharSequence, results: Filter.FilterResults) {
                mFilteredEvents = results.values as ArrayList<Event>
                notifyDataSetChanged()
            }
        }
    }
}
